lib/master_manipulator/service.rb

```diff
@@ -2,71 +2,222 @@
-      # if its PE and not 3.8 use the status endpoint
-      # it its not PE or it is 3.8 use the simple curl call
-      if pe_ver && !pe_ver.match(three_eight_regex)
-        curl_call = "-k -X GET -H 'Content-Type: application/json' https://#{hostname}:8140/status/v1/services?level=debug"
+    
+      if opts[:hup?]
+        hup_server(host, service_name)
+        return
       else
-        curl_call = "-I -k https://#{hostname}:8140"
+        on(host, puppet("resource service #{service_name} ensure=stopped"))
+        on(host, puppet("resource service #{service_name} ensure=running"))
       end
 
-      (1..opts[:wait_cycles]).each do |i|
-
-        @result = curl_on(host, curl_call, :acceptable_exit_codes => [0,1,7])
-        # parse body if we are using PE and we are not in PE 3.8
-        (pe_ver && !pe_ver.match(three_eight_regex)) ? @body = JSON.parse(@result.stdout) : @body = []
+      if opts[:is_pe?]
+        hostname = on(host, 'hostname').stdout.strip
+        opts[:wait_cycles] ||= 10
+        pe_ver = pe_version(host)
+        three_eight_regex = /^3\.8/
+        
+        # This logic is not ideal refactor in the future
+        # if its PE and not 3.8 use the status endpoint
+        # it its not PE or it is 3.8 use the simple curl call
+        if !pe_ver.match(three_eight_regex)
+          curl_call = "-I -k https://#{hostname}:8140"
+        else
+          curl_call = "-k -X GET -H 'Content-Type: application/json' https://#{hostname}:8140/status/v1/services?level=debug"
+        end
 
-        case @result.exit_code.to_s
+        (1..opts[:wait_cycles]).each do |i|
+          @result = curl_on(host, curl_call, :acceptable_exit_codes => [0,1,7])
+          # parse body if we are using PE and we are not in PE 3.8
+          (pe_ver && !pe_ver.match(three_eight_regex)) ? @body = JSON.parse(@result.stdout) : @body = []
+          
+          case @result.exit_code.to_s
           when '0'
             sleep 20
             pe_ver.match(/three_eight_regex/) ? return : (return if @body.all? { |k, v| v['state'] == 'running' })
           when '1', '7'
             # Exit code 7 is "connection refused"
             sleep (i**(1.2))
+          end
+        end
+
+        total_time = Time.now - start_time
+        
+        message = "Attempted to restart #{opts[:wait_cycles]} times, " +
+          "waited #{total_time} seconds total. \nHere is the status " +
+          "reported by the puppetserver:\n"
+        @body.each do |k,v|
+          message << "\n'#{k}' state: #{v['state']} "
         end
+        raise message
+      end
+    end
+
+    # Send SIGHUP to service_name on host and wait for reload to complete
+    # ==== Attributes
+    # *+host+ - the host that this should operate on
+    # *+service_name+ - the service to restart
+    #
+    # ==== Example
+    # hup_server(master, 'puppetserver')
+    # hup_server(host, 'pe-puppetserver')
+    def hup_server(host, service_name)
+      timeout = 30
```

Comment:
I anticipate future modifications to `#hup_server` such that it can HUP other services beside puppetserver or pe-puppetserver. I like the idea but I don't want to lock in on `host_hash['puppetservice']` as a default yet (I think that's what it is)